File: Graph/js/longestPath.js
const longestPath = (graph) => {
  const distance = {};
  for (let node in graph) {
    if (graph[node].length === 0) {
      distance[node] = 0;
    }
  }

  for (let node in graph) {
    traverseDistance(graph, node, distance);
  }

  return Math.max(...Object.values(distance));
};

const traverseDistance = (graph, node, distance) => {
  if (node in distance) return distance[node];

  let maxDistance = 0;
  for (let neighbor of graph[node]) {
    const attempt = traverseDistance(graph, neighbor, distance);
    if (attempt > maxDistance) maxDistance = attempt;
  }

  distance[node] = 1 + maxDistance;
  return distance[node];
};

File: Graph/go/longest_path.go
package graph

// LongestPath returns the length of the longest path in a directed acyclic graph.
func LongestPath(g Graph) int {
	distance := map[string]int{}
	for node, neighbors := range g {
		if len(neighbors) == 0 {
			distance[node] = 0
		}
	}

	maxDistance := 0
	for node := range g {
		current := traverseDistance(g, node, distance)
		if current > maxDistance {
			maxDistance = current
		}
	}
	return maxDistance
}

func traverseDistance(g Graph, node string, distance map[string]int) int {
	if val, ok := distance[node]; ok {
		return val
	}

	maxNeighbor := 0
	for _, neighbor := range g[node] {
		dist := traverseDistance(g, neighbor, distance)
		if dist > maxNeighbor {
			maxNeighbor = dist
		}
	}

	distance[node] = 1 + maxNeighbor
	return distance[node]
}
